book_source/source/2darrays.md

```diff
@@ -0,0 +1,237 @@
+array[3][4] = array[1][2];
+array[4][4] = 5;
+for (int i = 0; i < 5; ++i) {
+for (int j = 0; j < 5; ++j) {
+std::cout << array[i][j] << " ";
+}
+std::cout << std::endl; 
+}
+```
+````
+
+Similarily, if we want to change the value at (3,4) of the 2D array to 1, where 3 is the "x" coordinate or outside array and 4 is the "y" or inner array we would do something like array[3][4] = 1. Note that the 2D array can only be of one type, just like how a 1D array is only one type.
+
+## Looping over 2D Arrays
+
+We can also use a for loop to view and/or modify each index, although since this is a nested data structure (meaning there is a data structure within another data structure), we need to use a nested for loop in return. The starting value for each loop will typically start at 0, going to either array.length or array[0].length. Note the second for loop, we use array[0].length since we only want to loop through to the end of the inner array length, represented by array[0].
+
+
+````{tab-set-code}
+
+```{code-block} java
+int[][] array = new int[5][5];
+// Code to fill array with some values
+
+int sum = 0;
+for (int i = 0; i < array.length; i++) {
+    for (int j = 0; j < array[0].length; j++) {
+        sum += array[i][j];
+    }
+}
+```
+
+```{code-block} python
+array = [0] * 5  # Short-hand to make a length 5 list
+# Code to fill array with some values
+
+sum = 0
+for i in range(len(array)):
+    sum += array[i]
+```
+
+```{code-block} javascript
+let array = [0, 0, 0, 0, 0];
+// Code to fill array with some values
+
+let sum = 0;
+for (let i = 0; i < array.length; i++) {
+    sum += array[i];
+}
+```
+
+```{code-block} c
+int array[5];
+// Code to fill array with some values
+
+int sum = 0;
+size_t length = sizeof(array) / sizeof(array[0]);
+for (int i = 0; i < length; i++) {
+    sum += array[i];
+}
+```
+
+```{code-block} c++
+int array[5];
+// Code to fill array with some values
+
+int sum = 0;
+size_t length = sizeof(array) / sizeof(array[0]);
+for (int i = 0; i < length; i++) {
+    sum += array[i];
+}
+```
+
+```{code-block} c++
+int array[5];
+// Code to fill array with some values
+
+int sum = 0;
+for (int i = 0; i < array.length; i++) {
+    sum += array[i];
+}
+```
+````
+
+## Printing 2D Arrays
+
+Note that unlike with a 1D array, we cannot simply do Arrays.toString(array) on the entire 2D array, instead we have to loop thorugh each of the rows of the 2D array, then call Arrays.toString(array[i]), where i are values 0 through array.length - 1.
+
+````{tab-set-code}
+
+```{code-block} java
+
+int[][] array = new int[5][5];
+array[1][2] = 1;
+array[1][3] = 14;
+array[3][4] = array[1][2];
+array[4][4] = 5;
+
+for (int i = 0; i < array.length; i++) {
+    System.out.println(Arrays.toString(array[i]));
+}
+```
+
+```{code-block} python
+array = [0] * 3  # Short-hand to make a length 3 list
+array[0] = 8
+array[1] = 1
+array[2] = 7
+
+print(array)
+// Output: [8, 1, 7]
+```
+
+```{code-block} javascript
+let array = [[], [], [], []];
+
```

Comment:
Are we missing the step of actually putting values into the array?

Reply:
This is still unresolved @WateringFire